mmgen/models/architectures/sngan_proj/modules.py

```diff
@@ -0,0 +1,432 @@
+        norm_cfg (dict, optional): Config for normalization method. Default
+            to ``dict(type='BN')``.
+        auto_sync_bn (bool, optional): Whether convert Batch Norm to
+            Synchronized ones when Distributed training is on. Defualt to True.
+        eps (float, optional): eps for Normalization layers (both conditional
+            and non-conditional ones). Default to 1e-4.
+        init_method (str, optional): Initialization method for embedding
+            layers, 'ortho' | 'constant'. If sets as 'constant', embedding
+            layer for weight and bias are initialized as 1 and 0, respectively.
+            Default to 'ortho'.
+        reweight_embedding (bool, optional): (TODO: change name?) Whether add
+            ``1.`` to weight embedding result.  Default to True.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 num_classes,
+                 use_cbn=True,
+                 norm_cfg=dict(type='BN'),
+                 cbn_norm_affine=False,
+                 auto_sync_bn=True,
+                 eps=1e-4,
+                 init_method='ortho',
+                 reweight_embedding=True):
+        super().__init__()
+        self.use_cbn = use_cbn
+        self.reweight_embedding = reweight_embedding
+
+        norm_cfg = deepcopy(norm_cfg)
+        norm_type = norm_cfg['type']
+
+        if norm_type not in ['IN', 'BN', 'SyncBN']:
+            raise ValueError('Only support `IN` (InstanceNorm), '
+                             '`BN` (BatcnNorm) and `SyncBN` for '
+                             'Class-conditional bn. '
+                             f'Receive norm_type: {norm_type}')
+
+        if self.use_cbn:
+            norm_cfg.setdefault('affine', cbn_norm_affine)
+        norm_cfg.setdefault('eps', eps)
+        if check_dist_init() and auto_sync_bn and norm_type == 'BN':
+            norm_cfg['type'] = 'SyncBN'
+
+        _, self.norm = build_norm_layer(norm_cfg, in_channels)
+
+        if self.use_cbn:
+            if num_classes <= 0:
+                raise ValueError('`num_classes` must be larger '
+                                 'than 0 with `use_cbn=True`')
+            self.weight_embedding = nn.Embedding(num_classes, in_channels)
+            self.bias_embedding = nn.Embedding(num_classes, in_channels)
+
+        self.init_weights(init_method)
+
+    def forward(self, x, y=None):
+        """Forward function.
+
+        Args:
+            x (Tensor): Input tensor with shape (n, c, h, w).
+            y (Tensor, optional): Input label with shape (n, ).
+                Default None.
+
+        Returns:
+            Tensor: Forward results.
+        """
+
+        out = self.norm(x)
+        if self.use_cbn:
+            weight = self.weight_embedding(y)[:, :, None, None]
+            bias = self.bias_embedding(y)[:, :, None, None]
+            if self.reweight_embedding:
+                weight = weight + 1.
```

Comment:
Please check this implementation and specify the cases where we would adopt `self.reweight_embedding`.